spec/kinesis_client/kinesis_client_spec.rb

```diff
@@ -331,29 +278,52 @@ def error(*args)
           {
             data: "encoded 5",
             partition_key: "hashed"
-          },
+          }
         ],
         stream_name: 'fake-stream'
       }).and_return(@mock_failed_response)
+
+      expect($logger).to receive(:warn)
+      .with("Batch sent to fake-stream with 1 failures: [{:record=>{:data=>\"encoded 5\", :partition_key=>\"hashed\"}, :error_message=>\"error\"}]")
+    
       @kinesis_client << '4'
       @kinesis_client << '5'
       @kinesis_client.push_records
-      expect(@kinesis_client).to receive(:push_records)
-      @kinesis_client.retry_failed_records
     end
   end
 
-  describe "#decode_failed_records" do
-    it('decodes the records in @failed_records') do
-      allow(@mock_client).to receive(:put_records).and_return(@mock_failed_response)
-      @kinesis_client << '4'
-      @kinesis_client << '5'
-      @kinesis_client.push_records
-      @kinesis_client.decode_failed_records
-      expect(@kinesis_client.failed_records).to eql(['5'])
+  describe "#filter_failures" do
+    it "should return records and their error messages that do not enter the kinesis stream" do
+        @kinesis_client << '1'
+        @kinesis_client << '2'
+        batch = [{:data=>"encoded 1", :partition_key=>"hashed"},{:data=>"encoded 2", :partition_key=>"hashed"}]
+        expect(@kinesis_client
+          .filter_failures(@mock_failed_response, batch))
+          .to eql([{ :record=>{ :data=>"encoded 2", :partition_key=>"hashed" }, :error_message=>"error" }])
     end
   end
 
+  describe "#failed_records" do
+    it "should return decoded failed records" do
+      allow(@mock_client).to receive(:put_records).with({
+        records: [
+          {
+            data: "encoded 1",
+            partition_key: "hashed"
+          },
+          {
+            data: "encoded 2",
+            partition_key: "hashed"
+          }
+        ],
+        stream_name: 'fake-stream'
+      }).and_return(@mock_failed_response)
+      @kinesis_client << '1'
+      @kinesis_client << '2'
+      @kinesis_client.push_records
```

Comment:
Same comment as above. We should be able to just set `@failed_records` and then mock `avro.decode` to output what we're testing.